reconcile/unleash_feature_toggles/integration.py

```diff
@@ -0,0 +1,310 @@
+        instances = [
+            ui
+            for ui in unleash_instances_query(query_func).instances or []
+            if (not instance or ui.name == instance) and ui.admin_token
+        ]
+
+        # Set default values for all feature toggles
+        for inst in instances:
+            for project in inst.projects or []:
+                for feature_toggle in project.feature_toggles or []:
+                    feature_toggle.unleash.q_type = (
+                        feature_toggle.unleash.q_type or "release"
+                    )
+                    feature_toggle.unleash.impression_data = (
+                        feature_toggle.unleash.impression_data or False
+                    )
+        return instances
+
+    def fetch_current_state(self, client: UnleashServer) -> list[Project]:
+        """Fetch the current state of all Unleash projects including their feature toggles."""
+        return client.projects(include_feature_toggles=True)
+
+    def get_project_by_name(
+        self, projects: Iterable[UnleashProjectV1], name: str
+    ) -> UnleashProjectV1:
+        for project in projects:
+            if project.name.lower() == name.lower():
+                return project
+        raise ValueError(f"Project {name} not found")
+
+    def validate_unleash_projects(
+        self,
+        instance_name: str,
+        current_projects: Iterable[Project],
+        desired_projects: Iterable[UnleashProjectV1],
+    ) -> None:
+        """Validate that all projects referenced in the desired state are actually exist."""
+        current_project_names = {p.name.lower() for p in current_projects}
+        desired_project_names = {p.name.lower() for p in desired_projects}
+        if missing_projects := desired_project_names - current_project_names:
+            for p in missing_projects:
+                logging.error(f"[{instance_name}] Project '{p}' does not exist!")
+            raise ValueError(f"Non-existing projects: {missing_projects}")
+
+    def _reconcile_feature_toggles(
+        self,
+        client: UnleashServer,
+        instance: UnleashInstanceV1,
+        project_id: str,
+        dry_run: bool,
+        current_state: Sequence[FeatureToggle],
+        desired_state: Iterable[FeatureToggleUnleashV1],
+    ) -> None:
+        """Reconcile the feature toggles themselves."""
+        diff = diff_any_iterables(
+            current=current_state,
+            desired=[d for d in desired_state if d.delete is not True],
+            current_key=lambda c: c.name,
+            desired_key=lambda d: d.name,
+            equal=feature_toggle_equal,
+        )
+        for add in diff.add.values():
+            logging.info(
+                f"[{instance.name}/{project_id}] Adding feature toggle {add.name}"
+            )
+            try:
+                assert add.unleash.q_type  # make mypy happy
+                feature_type = FeatureToggleType[add.unleash.q_type]
```

Comment:
for kill switch type, `add.unleash.q_type` is `kill-switch`, this will fail as only `kill_switch` is accepted. `FeatureToggleType["kill-switch"]` will throw `KeyError: 'kill-switch'`

Reply:
Wow, nice catch! 🙇 Changed in https://github.com/app-sre/qontract-schemas/pull/650/commits/3a03ce9a4ad8cccaf58cae378abea00ab7130b11

Reply:
I haven't tested it in api call, does unleash api accept `kill-switch` or `kill_switch`

Reply:
The Unleash API takes `kill-switch` and in app-interface we use `kill_switch`. I've tested it.